### PR TITLE
adds caml_variant_hash port of c code

### PR DESCRIPTION
**Pull request by willemolding**
**Summary of changes**
Changes introduced in this pull request:
- Adds the `caml_hash_variant` function which replicates the hash algorithm used by OCaml to convert variant labels to strings
- Adds generic `Polyvar` struct which should serialize the same as a polyvar when used via bin-prot

**Reference issue to close (if applicable)**

Supporting work for #185  

**Other information and links**

### CODE REVIEW

**Review by hanabi1224 — APPROVED** [review]

File: protocol/bin-prot/src/polyvar/caml_hash_variant.rs [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+/* original C code
+    CAMLexport value caml_hash_variant(char const * tag)
+    {
+      value accu;
+      /* Same hashing algorithm as in ../typing/btype.ml, function hash_variant */
+      for (accu = Val_int(0); *tag != 0; tag++)
+        accu = Val_int(223 * Int_val(accu) + *((unsigned char *) tag));
+    #ifdef ARCH_SIXTYFOUR
+      accu = accu & Val_long(0x7FFFFFFFL);
```
Comment (hanabi1224) [inline protocol/bin-prot/src/polyvar/caml_hash_variant.rs:25]: It looks like the intermediate `accu` is `i64`, is it safe to `u32` in rust port? Maybe add some test cases where `u32` overflow on add happens

**Review by sudo-shashank — APPROVED** [review]

**Review by creativcoder — APPROVED** [review]

**Comment by willemolding** [comment]
Closing as this is included in #191
